mpi: implementing MPI sessions

hzhou opened the pull request:
## Pull Request Description
Implementing the MPI sessions from the 4.0 standards. 

This PR will only attempt to make a user program that uses sessions work. It does not separate the local initialization and global initialization, and it will only support `mpi://SELF` and `mpi://WORLD`. Missing the local/global internal separation, `MPI_Session_init` is not implemented as a local function but a collective instead. We will shift the global initialization to later communicator creation time in a separate PR. 

Reference: #4878

Split in #5016 #5013 #5015

[skip warnings]

## Expected Impact

This PR will make most sessions applications to work without altering ADI. The next PR will have to add new ADI and since it will split the `MPID_Init` functionality, without extra complications, the backward compatibility may get broken.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

minsii commented:
@hzhou Do I need to also review commits before `---- session commits start here ----`?

hzhou commented:
> @hzhou Do I need to also review commits before `---- session commits start here ----`?

Those are taken from PR #4941, not part of this PR.

minsii commented on src/mpi/init/mpir_init.c:
```diff
@@ -79,12 +79,18 @@ int MPIR_Init_impl(int *argc, char ***argv)
+
+int MPII_Init_thread(int *argc, char ***argv, int user_required, int *provided,
+                     MPIR_Session * session_ptr)
 {
```
It looks like the renaming of `MPIR_Init|Finalize` calls have been updated also in a previous commit
```commit 43de3764a9680b2bff46d8c96b709769bc5a2feb
Author: Hui Zhou <hzhou321@anl.gov>
Date:   Tue Dec 1 22:48:02 2020 -0600

    mpi/init: clean up init impl functions
```

Can you merge them?

minsii replied:
Besides, can you please explain why you define both `MPIR_Init_thread_impl` and `MPII_Init_thread`? It looks like `MPIR_Init_thread_impl` is not used anywhere. Same question to `MPIR_Finalize_impl`.

hzhou replied:
`MPIR_Init_thread_impl` and `MPIR_Finalize_impl` is called by `MPI-` functions (`MPI_Init_thread` and `MPI_Finalize`). The MPI functions are generated by the python binding scripts, thus you couldn't find them before `autogen`.

hzhou replied:
> 
> 
> It looks like the renaming of `MPIR_Init|Finalize` calls have been updated also in a previous commit
> 
> ```
> Author: Hui Zhou <hzhou321@anl.gov>
> Date:   Tue Dec 1 22:48:02 2020 -0600
> 
>     mpi/init: clean up init impl functions
> ```
> 
> Can you merge them?

I think you are referring to a commit in this PR https://github.com/pmodels/mpich/pull/4964 that is already merged.

minsii commented on test/mpi/init/session.c:
```diff
@@ -0,0 +1,156 @@
+    MPI_Comm_rank(lib_comm, &rank);
+
+    int sum;
+    MPI_Reduce(&rank, &sum, 1, MPI_INT, MPI_SUM, 0, lib_comm);
```
`sum` is not initialized. It may be set to a garbage value if `MPI_Reduce` does not work correctly.

hzhou replied:
That will be incorrect `MPI_Reduce` implementation and should be covered in one of the collective tests, right?

minsii replied:
It is a very small effort to make the test more robust (i.e., initialize sum) :-)

minsii commented on test/mpi/init/session.c:
```diff
@@ -0,0 +1,156 @@
+    int rank, size;
+
+    library_foo_init();
+
```
Should we do not continue if any error is found at `library_foo_init`?

hzhou replied:
This is the code I took from MPI standard text, so I would like to keep it as original as we can. Here, if `library_foo_init` fails, I think it should fail the later `MPI_Reduce` and fail the test. In practical use case, I agree that user should always check error return. That said, we have tons of testing code that ignores error return from MPI functions and I think they are okay for testing purposes. I am not arguing that we should not make our testing code more robust, but I think it is a quite big task and would like to leave it as separate tasks if it's ok.

minsii commented on test/mpi/init/session.c:
```diff
@@ -0,0 +1,156 @@
+    /*
+     * create a group from the WORLD process set
+     */
+    rc = MPI_Group_from_session_pset(lib_shandle, pset_name, &wgroup);
```
We don't use 3-line comment style for a single line comment. 
```
 /*
  * create a group from the WORLD process set
  */
```

Can we reformat it to  `/* create a group from the WORLD process set */`

hzhou replied:
Those are copied from the MPI standard text :). Anyway, they are fixed now.

minsii commented on test/mpi/init/session_psets.c:
```diff
@@ -0,0 +1,91 @@
+    if (rc != MPI_SUCCESS) {
+        MTestPrintfMsg(1, "Could not initialize session, bailing out\n");
+        return -1;
+    }
```
We need return `return MTestReturnValue(errs);` rather than `return -1;`

hzhou replied:
Fixed.

minsii commented on test/mpi/init/session_psets.c:
```diff
@@ -0,0 +1,91 @@
+        MPI_Session_get_nth_pset(shandle, MPI_INFO_NULL, i, &psetlen, NULL);
+        if (psetlen <= 0) {
+            printf("MPI_Session_get_nth_pset: returns psetlen = %d\n", psetlen);
+            errs++;
```
For error reporting, I think we should use `fprintf(stderr, ...)` but not `MTestPrintfMsg ` nor `printf `.

hzhou replied:
Actually, I am not sure what is the guideline. What is your reasoning?

hzhou replied:
I think I'll just use `printf`. MTest utilities simply prints to `stdout`, so I think it is consistent to just print to `stdout`. `MTestPrintfMsg` observes `verbosity` setting, so I guess all tests that doesn't have a lot of error messages can simply `printf`.

hzhou replied:
updated all tests to use `printf`

minsii replied:
I don't think we have any strict guideline for the error reporting. I just think it is a straightforward rule for programs to report any error on `stderr`. You are right that we are mixing the use of `printf` and `fprintf(stderr)` in many of our test programs. In fact, `MTestPrintErrorMsg` uses `printf` too. 

Anyway, I am OK if you prefer `printf` as long as it is consistent throughout the program.

minsii commented on test/mpi/init/session.c:
```diff
@@ -0,0 +1,156 @@
+    rc = MPI_Session_finalize(&lib_shandle);
+    if (rc != MPI_SUCCESS) {
+        MTestPrintfMsg(1, "MPI_Session_finalize returned %d\n", rc);
+        errs++;
```
For error reporting, I think we should use `fprintf(stderr, ...)`.

hzhou replied:
updated to all use `printf`

minsii commented on test/mpi/init/session_psets.c:
```diff
@@ -0,0 +1,91 @@
+        }
+
+        if (strstr(pset_name, use_pset_name) != NULL)
+            break;
```
Why do we use `strstr` but not `strcmp`?

minsii replied:
Can we use a single preallocated `pset_name[MPI_MAX_PSET_NAME_LEN]` to avoid calling `malloc`/`free` multiple times?

hzhou replied:
> 
> 
> Why do we use `strstr` but not `strcmp`?

From MPI standard text. User is likely to try the examples from standard, so I would like to keep this test very close to what is in the text.

hzhou replied:
> 
> 
> Can we use a single preallocated `pset_name[MPI_MAX_PSET_NAME_LEN]` to avoid calling `malloc`/`free` multiple times?

We could. But I would like to keep it as original as we can.

minsii commented on test/mpi/init/session.c:
```diff
@@ -0,0 +1,156 @@
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    library_foo_test();
+    MPI_Finalize();
+#else
```
For `MULT_INIT` test, can we also test calling two `MPI_Session_init`, and maybe need a multithreaded test too?

hzhou replied:
Yeah, probably a good idea

hzhou replied:
mult-thread test added.

minsii commented on src/mpi/group/group_impl.c:
```diff
@@ -622,3 +622,47 @@ int MPIR_Group_union_impl(MPIR_Group * group_ptr1, MPIR_Group * group_ptr2,
+    MPIR_Group *group_ptr;
+
+    if (strcmp(pset_name, "mpi://WORLD") == 0) {
+        mpi_errno = MPIR_Group_create(MPIR_Process.size, &group_ptr);
```
Is it possible to reuse `MPIR_Comm_group_impl` for `MPIR_Group_from_session_pset_impl`?

hzhou replied:
Most of the session API is local function that can't assume a communicator exist.

hzhou replied:
> When this function is called, I assume `MPIR_Process.comm_world` and `MPIR_Process.comm_self` are already established.

No, this assumption is false, so the checking of whether `comm_world` is established is critical. 

Now, in this PR, due to doing global init at `MPI_Session_init`, we always have `comm_world` active, But the code here cannot use that assumption. As a matter of fact, the next PR right after this will make the assumption no longer always true. The next PR will replace the assertions in the other branches with proper code as well.

The reason we do this PR separately from the next PR (proper local session init) is because this PR is still backward compatible; the next PR will have higher impact to downstream, thus better to separate.

hzhou replied:
> Besides,`MPIR_Comm_group_impl` is also a local function,

`MPIR_Comm_group_impl` is not a local function.

hzhou commented on maint/decode_handle:
```diff
@@ -25,10 +25,10 @@ my @kind_table = qw(
-    PROCGROUP
     VCONN
-    WORKQ_ELEM
     GREQ_CLASS
```
I believe both are already deleted in the main code (I think it was by you). We just neglected to update this utility tool.

hzhou commented on src/include/mpichconfconst.h:
```diff
@@ -60,11 +60,9 @@
-#define MPICH_MPI_STATE__IN_INIT 3      /* can call MPIR_Err_return_comm(...) on error */
+#define MPICH_MPI_STATE__UNINITIALIZED     0
+#define MPICH_MPI_STATE__LOCAL_INITIALIZED 1
+#define MPICH_MPI_STATE__INITIALIZED       2
```
Per MPI 4.0 standard, `MPI_Session_init` is a local function. This PR calls `MPII_Init_thread`, which is a global collective, thus is strictly incorrect. Although for any reasonable practical MPI application, all process will call `MPI_Sessioin_init` or one of the init anyway, so this PR probably will satisfy all practical usages. However, since it is strictly incorrect implementation, next PR will split the current "init" into strictly "local" init -- where we don't do any address exchange or communication, and a "global/collective" init -- where we do address exchange and creating communicators. `MPICH_MPI_STATE__LOCAL_INITIALIZED` is preparing for that.

hzhou replied:
On further thoughts, we actually don't need the `MPICH_MPI_STATE__LOCAL_INITIALIZED` to track local init state. We could just use `MPICH_MPI_STATE__INITIALIZED`. The world init state can and should be checked by whether `MPIR_Process.comm_world` is set -- I should enforce that in the later PRs.

However, we do need a state to signal (currently only the error utilities) that we finished mpr-layer init. I have renamed the state to `MPICH_MPI_STATE__MPIR_INITIALIZED`, I think that reflects the nature much better.

hzhou commented on src/include/mpichconfconst.h:
```diff
@@ -66,4 +66,9 @@
+#define MPICH_WORLD_MODEL_INITIALIZED   1
+#define MPICH_WORLD_MODEL_FINALIZED     2
+
 #endif /* MPICHCONFCONST_H_INCLUDED */
```
No. They are for different purposes. `mpich_state` is to track internal initialization state, to decide which parts of functionality is available. `MPIR_world_modle_state` is only used for `MPI_Initialized()` and `MPI_Finalized()`, which are "world modle" only functions. Under world model, we only can call `MPI_Init` and `MPI_Finalize` once, and `MPIR_world_model_state` is only used to track that. Before we kinda used `mpich_state` for both purpose. With session, we have to separately track them.

hzhou commented on src/mpi/comm/comm_impl.c:
```diff
@@ -623,6 +623,64 @@ int MPIR_Comm_create_group(MPIR_Comm * comm_ptr, MPIR_Group * group_ptr, int tag
+     *
+     * The next implementation will relax but assume the first call to this function will
+     * either from a "world" pset or "self" pset, during former a comm world will be created.
+     *
```
Next PR -- splitting init into local init and global init -- will be added before the alpha release. The final implementation will be added before beta. Sure, I'll mark them as TODO.

hzhou replied:
The "TODO"s are inside the branches in the following code where they are more relevant. I added reference comment and revised the TODO comments in the branches to make them more clear.

hzhou commented on src/mpi/comm/comm_impl.c:
```diff
@@ -623,6 +623,64 @@ int MPIR_Comm_create_group(MPIR_Comm * comm_ptr, MPIR_Group * group_ptr, int tag
+    HASH_VALUE(stringtag, n, hash);
+
+    int tag = hash % (MPIR_Process.attrs.tag_ub);
+
```
The interface uses a string so we have to convert this string into an integer some how. Hashing is one way to do that. This function potentially is used to establish first communicator, or it maybe used concurrently in the middle or in different threads, so I don't think there is better way to come up with tags other than hashing it.

Having a dedicated mask can help reduce the possibility of tag collision, so I probably will add that later. I probably will follow what you do in the partitioned communication implementation.

hzhou replied:
Added more specific comments.

hzhou replied:
It is only two lines and it is only used here, making it a separate routine seems unnecessary. Do I need improve the comments?

hzhou commented on src/mpi/comm/comm_impl.c:
```diff
@@ -623,6 +623,64 @@ int MPIR_Comm_create_group(MPIR_Comm * comm_ptr, MPIR_Group * group_ptr, int tag
+            mpi_errno = MPIR_Comm_group_impl(MPIR_Process.comm_world, &world_group_ptr);
+            MPIR_ERR_CHECK(mpi_errno);
+            MPIR_Group_release(world_group_ptr);
+        }
```
`MPI_Comm_create_group` is essentially creating a "sub" comm. `MPI_Comm_create_from_group`, on the other hand, is potentially creating a first communicator or a communicator that gathers arbitrary processes that only partially overlaps with existing comm. Remember, in a session model, it is not given that we have a all-to-all connection or even address exchange in place. So `MPI_Comm_create_from_group` potentially is the "global" initialization.

Here of course I took the easy way out by assuming the world comm is already established during `MPI_Session_init`, so it can be simply a `MPI_Comm_create_group` from the world comm. But the code is structured partially to prepare for the next PR where we may need do the `global` initialization here. 

If we *only* support `MPI://SELF` and `MPI://WORLD`, then we do (almost) nothing here with `MPI://SELF`, and we do *global* initialization if it is `MPI://WORLD`, or we do `MPI_Comm_create_group` here if world comm is already established (current implementation). 

In advanced implementation, where we prepare to support arbitrary proc set, then we'll have to do ad-hoc semi-dynamic address exchange and communicator creation -- pushed later down the road.

hzhou commented on src/mpi/comm/comm_impl.c:
```diff
@@ -623,6 +623,64 @@ int MPIR_Comm_create_group(MPIR_Comm * comm_ptr, MPIR_Group * group_ptr, int tag
+        /* TODO: create and dup MPIR_Process.comm_world */
+        MPIR_Assert(0 && "not implemented");
+        goto fn_fail;
+    } else if (group_ptr->pset_name && strcmp(group_ptr->pset_name, "mpi://SELF") == 0) {
```
The current code should always fall to `MPIR_Comm_create_group` since we always have the world comm created during `MPI_Session_init`. When user specify `mpi://SELF`, it'll be just group of 1 and it should work.

hzhou replied:
This PR will be kept as assuming `MPI_Session_init` will always be called collectively because this is the only way to keep backward compatibility for the device layer. I'll do separate PR to implement the local MPI_Session_init, at which point, other branches here will become valid possibilities (and will be implemented).

hzhou commented on src/mpi/init/mpir_finalize.c:
```diff
@@ -33,8 +33,7 @@ void MPIR_Add_finalize(int (*f) (void *), void *extra_data, int priority)
-        if (MPL_atomic_load_int(&MPIR_Process.mpich_state) != MPICH_MPI_STATE__PRE_INIT &&
-            MPL_atomic_load_int(&MPIR_Process.mpich_state) != MPICH_MPI_STATE__POST_FINALIZED) {
+        if (MPIR_Err_initialized()) {
             MPID_Abort(NULL, MPI_SUCCESS, 13, NULL);
```
You are looking at the code rather than the usage. The code checks for state, but the usage is for whether error utility is ready. I believe we should name the function according to its usage as it makes sense to the user. The code can change based on internal design. Here the connection between `MPICH_MPI_STATE__PRE_INIT` or `MPICH_MPI_STATE__POST_FINALIZED` and whether the error utility can be used is not obvious and in fact I would very much refactor it away. Adding a specific state variable such as `MPIR_err_is_initialized` is one such refactor.

hzhou replied:
I am marking this as resolved to clear my todo list. Feel free to unresolved them if needed.

hzhou replied:
Since you mentioned `MPIR_Err_uninitialized` in a later comment, I agree `MPIR_Err_initialized` is confusing. I have renamed it to `MPIR_Errutil_is_initialized` now, hopefully, this is clearer.

hzhou commented on src/mpi/init/mpir_finalize.c:
```diff
@@ -72,6 +72,10 @@ int MPIR_Finalize_impl(void)
+    if (MPIR_Process.init_counter > 0) {
+        goto fn_exit;
+    }
 
```
> If the user program makes mistake that num of MPI_Session_finalize calls > num of MPI_Session_init, ...

That is an error usage and we should detect and throw an error. Otherwise, the counter goes negative and it will mess up later `MPI_Session_init` -- this PR  can't deal with it but the standards allows it.

Your point is correct, and I'll add a error check for it. In fact, we probably should have a per-session counter to ensure the `MPI_Session_finalize` always matches the `MPI_Session_init`.

hzhou replied:
Added code in `MPIR_Session_finalize_impl` to free the session object. Because we have validation code validate the session handle, `MPI_Session_finaize` will be matched to `MPI_Session_init`. User wouldn't be able to `MPI_Session_finalize` the same session twice because the second time the session object will be invalid.

hzhou commented on src/mpi/init/mpir_init.c:
```diff
@@ -220,10 +222,9 @@ int MPIR_Init_thread_impl(int *argc, char ***argv, int user_required, int *provi
-    /* --BEGIN ERROR HANDLING-- */
-    return mpi_errno;
-    /* --END ERROR HANDLING-- */
+    goto fn_exit;
```
The comments marks the `return mpi_errno;` which is deleted, right?

hzhou reviewed (commented):
@minsii Thanks for the detailed review. I replied a few of your comments mainly to clarify the concepts. I'll do a full revision in January and ping you for review again.

hzhou commented on src/include/mpir_process.h:
```diff
@@ -35,7 +35,6 @@ extern MPIR_Object_alloc_t MPIR_Session_mem;
     MPL_atomic_int_t mpich_state;       /* Need use atomics due to MPI_Initialized() etc.
                                          * thread-safe per MPI-3.1.  See MPI-Forum ticket 357 */
-    MPIR_World_model_state world_model_state;
 
```
The previous commit that add `world_model_state` was a leftover.  Now removed.

hzhou commented:
Not done yet, but let make sure check the tests:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented on maint/decode_handle:
```diff
@@ -73,10 +73,10 @@ while (scalar @ARGV) {
+##   MPIR_GREQ_CLASS = 0xd,
+##   MPIR_SESSION    = 0xe,
+##   MPIR_INTERNAL   = 0xf,
 ##   } MPII_Object_kind;
```
I split the fix of `decode_handle` into a separate commit.

hzhou replied:
Let me move this commit to a separate PR. It is not essential for this PR.

hzhou commented on src/mpi/init/mpir_init.c:
```diff
@@ -91,7 +91,11 @@ int MPIR_Init_thread_impl(int *argc, char ***argv, int user_required, int *provi
+        *provided = MPIR_ThreadInfo.thread_provided;
+        goto fn_exit;
+    }
     /**********************************************************************/
```
The context get lost. There is no empty line changes here (any more). But I see there is a deleted empty line later above the `fn_fail` label, that is fixed now.

minsii replied:
line 106 in commit `init: add session_ptr to MPIR_Init_thread and MPIR_Finalize` is still a pure empty line change

hzhou commented on src/mpi/init/mpir_init.c:
```diff
@@ -208,6 +208,7 @@ int MPIR_Init_thread_impl(int *argc, char ***argv, int user_required, int *provi
 
     MPL_atomic_store_int(&MPIR_Process.mpich_state, MPICH_MPI_STATE__POST_INIT);
+    MPL_atomic_store_int(&MPIR_world_model_state, MPICH_WORLD_MODEL_INITIALIZED);
 
```
Addressed in https://github.com/pmodels/mpich/pull/4949#discussion_r545439540

hzhou commented on src/mpi/init/globals.c:
```diff
@@ -7,10 +7,11 @@
 
-MPIR_Process_t MPIR_Process = { MPL_ATOMIC_INT_T_INITIALIZER(MPICH_MPI_STATE__PRE_INIT) };
+MPIR_Process_t MPIR_Process;
 
```
Separated.

hzhou commented on src/mpi/errhan/errutil.c:
```diff
@@ -210,14 +210,8 @@ void MPII_Errhandler_set_fc(MPI_Errhandler errhand)
 /* --BEGIN ERROR HANDLING-- */
 void MPIR_Err_preOrPostInit(void)
 {
```
On that, my previous commit added `MPIR_Err_initialized` probably is very confusing, so I now renamed it to `MPIR_Errutil_is_initialized`. (need add a link to previous comments).

I don't think the name `MPIR_Err_preOrPostInit` is unclear, and folks may be more used to it. So I think we can leave this name for now.

hzhou commented on src/mpi/errhan/errutil.c:
```diff
@@ -226,8 +220,8 @@ void MPIR_Err_preOrPostInit(void)
-    return (state != MPICH_MPI_STATE__PRE_INIT && state != MPICH_MPI_STATE__POST_FINALIZED);
+    int state = MPL_atomic_load_int(&MPIR_mpich_state);
+    return (state != MPICH_MPI_STATE__UNINITIALIZED);
 }
```
`MPICH_MPI_STATE__LOCAL_INITIALIZED` (now renamed to `MPICH_MPI_STATE__MPIR_INITIALIZED`) is a transition state inside `MPIR_Init` and only covers error routines used by `MPID_Init`. Here `MPIR_Err_return_comm` is used by other `MPI-`layer functions that should never see this transition state.

hzhou commented on src/mpi/errhan/errutil.c:
```diff
@@ -249,8 +243,8 @@ int MPIR_Err_return_comm(MPIR_Comm * comm_ptr, const char fcname[], int errcode)
-        MPL_atomic_load_int(&MPIR_Process.mpich_state) == MPICH_MPI_STATE__POST_FINALIZED) {
+    int state = MPL_atomic_load_int(&MPIR_mpich_state);
+    if (state == MPICH_MPI_STATE__UNINITIALIZED) {
         /* for whatever reason, we aren't initialized (perhaps error
```
MPICH_MPI_STATE__LOCAL_INITIALIZED (now renamed to MPICH_MPI_STATE__MPIR_INITIALIZED) is a transition state inside MPIR_Init and only covers error routines used by MPID_Init. Here MPIR_Err_return_comm is used by other MPI-layer functions that should never see this transition state.

hzhou replied:
Added new commit to rewrite that condition using `!MPIR_Errutil_is_initialized()`.

hzhou commented on src/mpi/init/mpir_finalize.c:
```diff
@@ -126,8 +138,11 @@ int MPIR_Finalize_impl(void)
     MPII_thread_mutex_destroy();
     MPIR_Typerep_finalize();
+
     MPL_atomic_store_int(&MPIR_mpich_state, MPICH_MPI_STATE__UNINITIALIZED);
```
Fixed (in commit "init: add session_ptr to MPIR_Init_thread and MPIR_Finalize")

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
@minsii I think I addressed most of your comments. What left are the error reporting style in the tests and potentially adding a multi-threading test. Because there are many comments to keep track and the commits have been updated quite a few times from when the original comments were made, could you give it another round to review?

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

minsii commented:
Thanks. I will try to do review tonight or tomorrow.

hzhou commented:
Added multi-thread session test.

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
Looks like we can't allow a new session after the final "finalize" yet. Modified the test:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
The multi-threading failure is due to info utility cannot be used outside init/finalize -- until https://github.com/pmodels/mpich/pull/4978 gets merged. I updated the test and removed the info usage since it is not the scope of the test.

test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
@minsii The tests are clean. Ready for final review

minsii commented:
Thanks. My review was delayed due to some urgent tasks on my side. I will look into it asap

minsii commented on src/mpi/init/mpir_finalize.c:
```diff
@@ -66,11 +66,23 @@ static void MPIR_Call_finalize_callbacks(int min_prio, int max_prio)
+        /* handle any clean up on session */
+        MPIR_Handle_obj_free(&MPIR_Session_mem, session_ptr);
+    }
+
```
Since we have already defined `is_world_model`, can we check `if(! is_world_model )` instead of `if(session_ptr)`?

minsii commented on src/mpi/init/mpir_init.c:
```diff
@@ -221,6 +233,9 @@ int MPIR_Init_thread_impl(int *argc, char ***argv, int user_required, int *provi
+    if (!p_session_ptr) {
+        MPL_atomic_store_int(&MPIR_world_model_state, MPICH_WORLD_MODEL_INITIALIZED);
+    }
     MPIR_INIT_UNLOCK;
```
Any reason why we move state check of `MPIR_world_model_state` from line 204 to `fn_exit?`

minsii commented on test/mpi/init/session.c:
```diff
@@ -0,0 +1,158 @@
+
+    MPI_Info_get(tinfo, mt_key, sizeof(out_value), out_value, &flag);
+    if (flag != 1) {
+        printf("Could not find key %s\n", mt_key);
```
Standard defines the value of `flag` returned by `MPI_Info_get` to be `true` or `false`. Thus, better to use `if(!flag)` rather than `if(flag != 1)`, as the test may be used by different MPI implementations.

minsii commented on test/mpi/init/session_psets.c:
```diff
@@ -0,0 +1,92 @@
+#include "mpitest.h"
+
+/* This is adapted example 10.9 from MPI Standard 4.0 */
+
```
Can we also copy the brief description of the example from standard and add into the header comment?

minsii commented on test/mpi/init/session_psets.c:
```diff
@@ -0,0 +1,92 @@
+    rc = MPI_Group_from_session_pset(shandle, pset_name, &pgroup);
+    if (rc != MPI_SUCCESS) {
+        printf("Could not create grep from pset %s\n", pset_name);
+        errs++;
```
Is it typo "grep"->"group"?

minsii commented on test/mpi/init/session_psets.c:
```diff
@@ -0,0 +1,92 @@
+    } else {
+        printf("%d Errors\n", errs);
+    }
+    return MTestReturnValue(errs);
```
Can we use `MTest_Finalize(errs)` here to print the final message?

minsii commented on test/mpi/threads/init/mult_session.c:
```diff
@@ -0,0 +1,150 @@
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
```
I understand that you want to keep the test as close as to the original example. But all the tests check only `WORLD` pset. Can we extend the tests to also cover `SELF` pset?

minsii reviewed (changes requested):
Sorry again for the delay of my review. Please find inlined comments.

hzhou commented on src/include/mpir_process.h:
```diff
@@ -80,6 +80,8 @@ typedef struct MPIR_Process_t {
 extern MPIR_Process_t MPIR_Process;
 
+extern MPL_atomic_int_t MPIR_world_model_state;
+
```
Maybe I can turn some of them into local static variables...

hzhou commented:
Replaced by #5025
